proposals/0000-linear-types.rst

```diff
@@ -903,49 +903,86 @@ cope with affine functions. Therefore the choice between these three
+  So abandoning the stronger guarantee of linear types would come at
+  a cost for API designers.
+* While affine types are sufficiently strong to achieve many desirable
+  properties, linear types can express them just as well at minimal
+  implementation and API design cost.
+
+An example of a direct style API that crucially relies on linearity is
+@gelisam's `3D-printable models
 <https://www.spiria.com/en/blog/desktop-software/making-non-manifold-models-unrepresentable>`_).
-It would seem an arbitrary limitation for little gain to prevent them
-from using the linear types that they need. Plus even in ``RIO`` code,
-where exceptions do matter, linear types are useful: they allow prompt
-deallocation as argued in the section on Exceptions_, it can be much
-harder to reason on the lifetime of resources with explicit scopes
-like with ``bracket`` (see the `inline-java use-case
-<http://www.tweag.io/posts/2017-11-29-linear-jvm.html>`_ for an
-example where scopes have proved to be unsatisfactory).
-
-There is, nonetheless, value to affine types. There are some
-applications where affine types are enough to enforce invariants (such
-as in-place mutation of garbage-collected structure, like mutable
-arrays). And they can presumably benefit from the additional
-flexibility. For instance, ``catch`` can get a more fine-grained type
-(writing ``'A`` for the affine multiplicity):
+Exceptions can only be caught in the ``IO`` monad, yet this API is
+pure. So exceptions are not a concern in the design of this API. The
+properties this API wants to enforce hold even with linear types and
+even in the face of exceptions being thrown (in a pure or impure
+context) and caught (in an impure context). No linear types means this
+API would need to use CPS, if that works at all to enforce the same
+properties.
+
+Another example is `language interop
+<http://www.tweag.io/posts/2017-11-29-linear-jvm.html>`_ by
+@facundominguez and @mboes. In this example, Haskell users create GC
+roots for every object in the JVM's heap that they want to reference
+directly. These GC roots must be released as soon as the reference is
+no longer useful, otherwise memory pressure on the JVM will increase
+and throughput will suffer because of more frequent GC passes. Worse,
+deleting roots too late can sometimes prevent any memory from being
+reclaimed at all, leading to out-of-memory conditions. Introducing
+a ``bracket``-like ``withJvmScope`` action is one way to ensure all
+roots do get deleted eventually (at scope exit), but in practice, in
+complex dual-language projects, introducing neither too fine-grained
+or too coarse-grained scopes has proven very difficult. Furthermore,
+``bracket``-like constructs break tail-recursion. Linear types enable
+working with a single global resource scope, while still guaranteeing
+eventual deletion of roots, in any order. Affine types do not. At any
+rate, not in direct-style.
```

Comment:
This paragraph is a tad long. Maybe we ought to par down the most technical details, and defer them to the blog post (following the mantra that the blog post has more space and spends quite a bit of space and effort into making the explanation clear).

Reply:
We could, but my concern is that the blog post actually does not explain in a single pithy paragraph why multiple scopes are so bad. If the blog post did have that, then I'd be happy to link to that instead. That said, I found a couple of sentences to cut out, so I'll do that.